src/plugins/git/porcelain.js

```diff
@@ -0,0 +1,264 @@
+    if (n.type() !== "TREE") {
+      throw new Error(`Unable to cast ${n.type()} to Tree`);
+    }
+    return new Tree(n.graph, n.nodeAddress);
+  }
+
+  hash(): Hash {
+    return this.address().id;
+  }
+
+  entries(): TreeEntry[] {
+    return this.graph
+      .neighborhood(this.nodeAddress, {
+        nodeType: "TREE_ENTRY",
+        edgeType: "INCLUDES",
+        direction: "OUT",
+      })
+      .map(({neighbor}) => new TreeEntry(this.graph, neighbor));
+  }
+
+  entry(name: string): ?TreeEntry {
+    return this.entries().filter((te) => te.name() === name)[0];
+  }
+}
+
+export class TreeEntry extends BaseGitPorcelain {
+  static from(n: BaseGitPorcelain): TreeEntry {
+    if (n.type() !== "TREE_ENTRY") {
+      throw new Error(`Unable to cast ${n.type()} to TreeEntry`);
+    }
+    return new TreeEntry(n.graph, n.nodeAddress);
+  }
+
+  name(): string {
+    const node: Node<TreeEntryNodePayload> = (this.node(): any);
+    return node.payload.name;
+  }
+
+  evolvesTo(): TreeEntry[] {
+    return this.graph
+      .neighborhood(this.nodeAddress, {
+        nodeType: "TREE_ENTRY",
+        edgeType: "BECOMES",
+        direction: "OUT",
+      })
+      .map(({neighbor}) => new TreeEntry(this.graph, neighbor));
+  }
+
+  evolvesFrom(): TreeEntry[] {
+    return this.graph
+      .neighborhood(this.nodeAddress, {
+        nodeType: "TREE_ENTRY",
+        edgeType: "BECOMES",
+        direction: "IN",
+      })
+      .map(({neighbor}) => new TreeEntry(this.graph, neighbor));
+  }
+
+  contents(): Tree | Blob | SubmoduleCommit {
+    const neighbors = this.graph
+      .neighborhood(this.nodeAddress, {
+        edgeType: "HAS_CONTENTS",
+        direction: "OUT",
+      })
+      .map(({neighbor}) => neighbor);
+    if (neighbors.length !== 1) {
+      throw new Error(
+        `Wrong number of contents for ${stringify(this.address())}`
+      );
+    }
+    const n = neighbors[0];
+    const type = n.type;
+    switch (type) {
+      case "BLOB":
+        return new Blob(this.graph, n);
+      case "TREE":
+        return new Tree(this.graph, n);
+      case "SUBMODULE_COMMIT":
+        return new SubmoduleCommit(this.graph, n);
+      default:
+        throw new Error(`Neighbor had invalid type ${type}`);
+    }
+  }
+}
+
+export class Blob extends BaseGitPorcelain {
+  static from(n: BaseGitPorcelain): Blob {
+    if (n.type() !== "BLOB") {
+      throw new Error(`Unable to cast ${n.type()} to Blob`);
+    }
+    return new Blob(n.graph, n.nodeAddress);
+  }
+
+  hash(): Hash {
+    return this.nodeAddress.id;
+  }
+}
+
+export class SubmoduleCommit extends BaseGitPorcelain {
+  static from(n: BaseGitPorcelain): SubmoduleCommit {
+    if (n.type() !== "SUBMODULE_COMMIT") {
+      throw new Error(`Unable to cast ${n.type()} to SubmoduleCommit`);
+    }
+    return new SubmoduleCommit(n.graph, n.nodeAddress);
+  }
+
+  url(): string {
+    const node: Node<SubmoduleCommitPayload> = (this.node(): any);
```

Comment:
All these `any`s are disturbing. Let's parameterize `BaseNodePorcelain` over a node payload, as in the GitHub plugin.